core/offchain/src/lib.rs

```diff
@@ -60,9 +60,17 @@ pub mod testing;
 pub use offchain_primitives::OffchainWorkerApi;
 
 /// Provides currently configured authority key.
-pub trait AuthorityKeyProvider: Clone + 'static {
+pub trait AuthorityKeyProvider<Block: traits::Block>: Clone + 'static {
+	/// The crypto used by the block authoring algorithm.
+	type ConsensusPair: crypto::Pair;
+	/// The crypto used by the finality gadget.
+	type FinalityPair: crypto::Pair;
+
 	/// Returns currently configured authority key.
-	fn authority_key<TPair: crypto::Pair>(&self) -> Option<TPair>;
+	fn authority_key(&self, block_id: &BlockId<Block>) -> Option<Self::ConsensusPair>;
```

Comment:
it's no worse than at present, so i'll let it pass, but in general i'd prefer to avoid any notion of "authority key" over this API. there should really just be the key store and if you want the Babe, Aura or Grandpa key, then you have to consult that module should implement/use a special off-chain API that allows you to find it.